File: Yr2022/day_fifteen.py
from time import perf_counter_ns
import numpy as np
from dataclasses import dataclass
from typing import Tuple, List
import itertools
from tqdm import tqdm


def sign(x):
    return -1 if x < 0 else 1 if x>0 else 0

class Sensor:
    def __init__(self, pos: Tuple[int, int], nrst_b: Tuple[int, int]):
        self.pos = pos
        self.nrst_b = nrst_b
        self.taxi = self.taxicab_distance


    @property
    def x(self):
        return self.pos[0]

    @property
    def y(self):
        return self.pos[1]

    @property
    def taxicab_distance(self):
        delta_x, delta_y = self.nrst_b[0]-self.pos[0], self.nrst_b[1]-self.pos[1]
        return abs(delta_x) + abs(delta_y)

    def range_from(self, test_pos):
        delta_x, delta_y = test_pos[0]-self.pos[0], test_pos[1]-self.pos[1]
        return abs(delta_x) + abs(delta_y)

    def in_range(self, test_pos):
         return self.range_from(test_pos)<= self.taxi


    def annulus(self):
        result = set()
        for dx in range(0, self.taxi+2):
            dy = self.taxi + 1 - dx
            result.add((self.pos[0]+dx, self.pos[1]+dy))
            ### Don't need to generate these as can just check NE and SW sectors
            #result.add((self.pos[0]-dx, self.pos[1]+dy))
            #result.add((self.pos[0]+dx, self.pos[1]-dy))
            result.add((self.pos[0]-dx, self.pos[1]-dy))
        return result
    

    def get_conjugate_coord(self, coord, is_min=True, is_y=True):
        dist = self.pos[1] - coord if is_y else self.pos[0] - coord
        conj_dist = abs(dist)-self.taxi if is_min else self.taxi - abs(dist)
        return self.pos[0] + conj_dist if is_y else self.pos[1] + conj_dist


@dataclass
class SensorList:
    sensors: List[Sensor]

    def is_sensor_in_range(self, test_pos):
        for sensor in self.sensors:
            if sensor.in_range(test_pos):
                return True
        return False


    def is_beacon(self, test_pos):
        for sensor in self.sensors:
            if test_pos == sensor.nrst_b:
                return True
        return False


    def annulus_set(self):
        result = set()
        result_list = []
        for sensor in tqdm(self.sensors):
            result_list.append(sensor.annulus())        
        for pair in tqdm(list(itertools.combinations(result_list, 2))):
            onions = pair[0].intersection(pair[1])
            result = result.union(onions)
        return result
          

    def check_row(self, row_idx, speed_mode=True):
        min_x = min([sens.get_conjugate_coord(row_idx) for sens in self.sensors if abs(sens.pos[1]-row_idx) <= sens.taxi ])
        max_x = max([sens.get_conjugate_coord(row_idx, is_min=False) for sens in self.sensors if abs(sens.pos[1]-row_idx) <= sens.taxi  ])
        if speed_mode:
            return max_x - min_x
        in_range_counter = 0
        for x in tqdm(range(min_x, max_x+1)):
            if self.is_sensor_in_range((x, row_idx)) and not self.is_beacon((x, row_idx)):
                in_range_counter += 1
        return in_range_counter


    def search_space(self, big):         
        min_x, max_x = 0, big
        min_y, max_y = 0, big 
        for (x,y) in tqdm(self.annulus_set()):
            if min_x<=x<=max_x and min_y<=y<=max_y:
                if not self.is_sensor_in_range((x,y)):
                    return x*big + y                    
        return None

    def search_space_better(self, big):
        ascend, descend = set(), set()
        for (a, b) in itertools.combinations(self.sensors, 2):            
            if a.range_from(b.pos) == a.taxi+b.taxi+2:
                print(a.pos, b.pos)
                if sign(a.x-b.x) == sign(a.y-b.y):
                    a_sum, b_sum = a.x + a.y, b.x + b.y
                    descend.add(a_sum + sign(b_sum-a_sum)*(a.taxi+1))
                else:
                    a_dif, b_dif = a.y - a.x, b.y - b.x
                    ascend.add(a_dif + sign(b_dif-a_dif)*(a.taxi+1))
        for (up, dwn) in itertools.product(ascend, descend):
            if (up % 2 == 0) ^ (dwn % 2 == 0):
                continue
            x,y = (dwn-up)/2, (dwn+up)/2
            for sensor in self.sensors:
                if sensor.range_from((x,y)) > sensor.taxi:
                    return int(x*big + y)
        return None



        

def find_beacons(filename: str):
    cmd_lines= []
    with open(filename, 'r') as file:
        cmd_lines = file.readlines()


    sensor_list = SensorList([])
    for line in cmd_lines:
        split_line = line.strip().replace('=',',').replace(':',',').split(',')
        sens_x, sens_y = int(split_line[1]), int(split_line[3])
        b_x, b_y = int(split_line[-3]), int(split_line[-1])
        sensor_list.sensors.append( Sensor((sens_x, sens_y), (b_x,b_y)))
    
    
    row_idx = 2_000_000
    part_one = sensor_list.check_row(row_idx, speed_mode=True)   
    #part_two = sensor_list.search_space(big = 4_000_000)    
    part_two = sensor_list.search_space_better(big = 4_000_000)

    return part_one, part_two
  

if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_fifteen.txt"     
    part_one, part_two = find_beacons(file_loc)
    
    print(f"(Part One): {part_one}     (Part Two): {part_two}")   
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")  

File: Yr2022/src/day_fifteen.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::time::Instant;
use std::collections::HashSet;
use itertools::Itertools;


fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_fifteen.txt";
    let (part_one, part_two) = search_beacons(file_name);
    let duration = start.elapsed();    
    println!("Part One: {}  Part Two: {}", part_one, part_two);
    println!("Time taken to execute: {} seconds", duration.as_secs());    
}

type Pos = (i64,i64);

struct Sensor {
    own_pos: Pos,
    nrst_b: Pos,
    taxi: i64,
}


impl Sensor {
    pub fn set_taxicab_distance(&mut self) {
        let delta_x = self.nrst_b.0-self.own_pos.0;
        let delta_y = self.nrst_b.1-self.own_pos.1;
        self.taxi = delta_x.abs() + delta_y.abs();
    }

    pub fn in_range(&self, test_pos: Pos) -> bool {
        let delta_x = test_pos.0 - self.own_pos.0;
        let delta_y = test_pos.1 - self.own_pos.1;
        return (delta_x.abs() + delta_y.abs()) <= self.taxi;
    }

    pub fn annulus(&self) -> HashSet<Pos> {
        let mut result : HashSet<Pos> = HashSet::new();
        for dx in 0..=self.taxi+1 {
            let dy = self.taxi + 1 - dx;
            result.insert( (self.own_pos.0+dx, self.own_pos.1+dy));
            // Don't need to include these
            //result.insert( (self.own_pos.0-dx, self.own_pos.1+dy));
            //result.insert( (self.own_pos.0+dx, self.own_pos.1-dy));
            result.insert( (self.own_pos.0-dx, self.own_pos.1-dy));
        }
        return result;
    }

    pub fn get_conj_coord(&self, coord: i64, is_min: bool, is_y : bool) -> i64 {        
        let dist = if is_y {self.own_pos.1 - coord} else { self.own_pos.0 - coord};
        let conj_dist = if is_min {dist.abs() - self.taxi as i64} else {self.taxi as i64 - dist.abs()};
        let result = if is_y {self.own_pos.0 + conj_dist} else {self.own_pos.1 + conj_dist};
        return result;
    }
}


struct SensorList {
    sensors: Vec<Sensor>,
}

impl SensorList {
    pub fn prep_taxi(&mut self) {
        self.sensors.iter_mut().for_each(|s| s.set_taxicab_distance());
    }

    pub fn is_sensor_in_range(&self, test_pos: Pos) -> bool {
        return self.sensors.iter().any(|s| s.in_range(test_pos)) ;
    }

    pub fn is_beacon(&self, test_pos: Pos) -> bool {
        return self.sensors.iter().any(|s| s.nrst_b == test_pos) ;
    }

    pub fn annulus_set(&self) -> HashSet<Pos> {
        let mut result : HashSet<Pos> = HashSet::new();
        let annulus_vec: Vec<HashSet<Pos>> = self.sensors.iter().map(|s| s.annulus() ).collect();
        for (annul_one, annul_two) in annulus_vec.iter().tuple_combinations() {
            result.extend(annul_one.intersection(&annul_two));
        }
        return result;
    }

    pub fn check_row(&self, row_idx: i64, speed_mode:bool) -> usize{
        let min_x: i64 = self.sensors.iter().map(|s| s.get_conj_coord(row_idx, true, true)).min().unwrap();
        let max_x: i64 = self.sensors.iter().map(|s| s.get_conj_coord(row_idx, false, true)).max().unwrap();
        if speed_mode {
            return (max_x - min_x) as usize;
        }
        let mut in_range_counter: usize = 0;
        for x in min_x..=max_x {
            if self.is_sensor_in_range((x, row_idx)) && !self.is_beacon((x, row_idx)) {
                in_range_counter += 1;
            }
        }
        return in_range_counter;
    }

    pub fn search_space(&self, big: i64) -> Option<i64> {
        let (min_x, max_x) : Pos = (0, big);
        let (min_y, max_y) : Pos = (0, big);
        let annulus_set : HashSet<Pos> = self.annulus_set();
        for pos in annulus_set {
            if (min_x..=max_x).contains(&pos.0) && (min_y..=max_y).contains(&pos.1) {
                if !self.is_sensor_in_range(pos) {
                    return Some( pos.0*big + pos.1 );
                }
            }
        }
        return None;
    }
}


fn search_beacons(my_file_name: &str) -> (usize, i64) {
    let file = File::open(my_file_name).expect("File does not exist");
    let buf_lines = BufReader::new(file).lines();
    let line_vec: Vec<String> = buf_lines.map(|l| l.expect("Fail to parse")).collect();

    let mut sensor_list : SensorList = SensorList {sensors: Vec::new()};
    for line in line_vec {
        let split_line: Vec<&str> = line.trim().split(&[',','=',':']).collect();
        let sens_x: i64 = split_line[1].parse::<i64>().unwrap();
        let sens_y: i64 = split_line[3].parse::<i64>().unwrap();
        let b_x: i64 = split_line[split_line.len() - 3].parse::<i64>().unwrap();
        let b_y: i64 = split_line[split_line.len() - 1].parse::<i64>().unwrap();
        sensor_list.sensors.push( Sensor { own_pos:(sens_x, sens_y), nrst_b:(b_x, b_y), taxi:0});
    }

    sensor_list.prep_taxi();
    let row_idx: i64 = 2000000;
    let part_one = sensor_list.check_row(row_idx, true);
    let mut part_two : i64 = 0 ;
    let part_two_result = sensor_list.search_space(4000000);
    match part_two_result {
        Some(val) => part_two = val,
        None => println!("Couldn't find a solution!"),
    }    
    return (part_one,part_two)
}
